### DISCUSSION TITLE
how to do multi cursor like vs code?

### DISCUSSION THREAD

**Discussion by cStralpt** [LazyVim/LazyVim#843] [Q&A] [answered]
i have installed mg979/vim-visual-multi , and it's weirdly conflict with lazyvim, like jumping somewhere and switching from normal to command mode constantly back and forward

**Comment by dpetka2001** [comment]
Neovim doesn't have any support for multi cursors as it doesn't really align with the vim philosophy. `vim-visual-multi` is your best bet, except if some other plugin provides this that I'm not aware of.

**Comment by cStralpt** [comment]
Any plans to add it to lazyvim as a feeature?

**Reply by dpetka2001** [reply, marked as answer]
The plugin `vim-visual-multi` you mean? You can just add that to your configuration and use it. Support for multi-cursor is not something related to Lazyvim. It would either have to be supported by Neovim natively (cuz Lazyvim is just a distro that uses Neovim) or someone would have to make a plugin that supports that (just like `vim-visual-multi`).

**Reply by cStralpt** [reply]
thank you for asnwering, after digging into some articles on internet, i think i dont need a multi cursor in neovim, i just sets this keymaps for insert modes and it can do what mostly multi cursor can doo

```
map("i", "<C-d>", function()
  local new_text = vim.fn.input("Replace with?: ")
  local cmd = "normal! *Ncgn" .. new_text
  vim.cmd(cmd)
end, { desc = "ctrl+d vs code alternative" })

```

**Reply by dpetka2001** [reply]
I think your own answer should be the answer for this topic. You're the one who provided a solution for yourself. I didn't answer anything. :stuck_out_tongue:

**Comment by coderup2021** [comment]
> thank you for asnwering, after digging into some articles on internet, i think i dont need a multi cursor in neovim, i just sets this keymaps for insert modes and it can do what mostly multi cursor can doo
> 
> ```
> map("i", "<C-d>", function()
>   local new_text = vim.fn.input("Replace with?: ")
>   local cmd = "normal! *Ncgn" .. new_text
>   vim.cmd(cmd)
> end, { desc = "ctrl+d vs code alternative" })
> ```

hello, where you put the config?  ~/.config/nvim/lua/config/keymaps.lua ?  I tried it, it don't work

**Reply by timjackleus** [reply]
It's because `map` probably is a local variable, to get this to work in your case replace it with `vim.keymap.set` e.g.

```
vim.keymap.set("i", "<C-d>", function()
  local new_text = vim.fn.input("Replace with?: ")
  local cmd = "normal! *Ncgn" .. new_text
  vim.cmd(cmd)
end, { desc = "ctrl+d vs code alternative" })
```
Also see `:help keymap.set` to get some more info on what all the arguments mean :)

**Comment by bazizi** [comment]
If you want to, let's say comment multiple lines at once (using `//`'s) you can do this (works in vanilla Vim as well):
- `<c-v>` to enter visual block mode
- Press `j` multiple times so the desired lines are selected
- Press `I` so you enter insert mode at the beginning of the lines
- Type `//`
- Press `<c-[>` (or `Esc`) to exit insert mode

**Reply by gybo03** [reply]
this solved all my problems thanks for answering

**Reply by LingLambda** [reply]
> If you want to, let's say comment multiple lines at once (using `//`'s) you can do this (works in vanilla Vim as well):
> 
>     * `<c-v>` to enter visual block mode
> 
>     * Press `j` multiple times so the desired lines are selected
> 
>     * Press `I` so you enter insert mode at the beginning of the lines
> 
>     * Type `//`
> 
>     * Press `<c-[>` (or `Esc`) to exit insert mode

Excuse me, Is this included in LazyVim by default, or do I need to install some additional plugins?

**Reply by gybo03** [reply]
i think this is the core functionality of Vim so by transitivity it is in LazyVim also, you dont need to install anything

**Reply by LingLambda** [reply]
> i think this is the core functionality of Vim so by transitivity it is in LazyVim also, you dont need to install anything

Strangely, I can't do it. When I press I, there's only one line with a cursor

**Reply by knokhen** [reply]
I think the problem is Lazyvim mapping <c-v> to pasting text like <c-c> for copy and <c-v> for past which conflicts with the default mapping.

**Comment by gybo03** [comment]
Yes it looks like it but if you make a change in that line and then press Esc to escape insert mode it will apply to all previous selected lines

On November 24, 2024 12:57:14 PM GMT+01:00, ling ***@***.***> wrote:
>> i think this is the core functionality of Vim so by transitivity it is in LazyVim also, you dont need to install anything
>
>Strangely, I can't do it. When I press I, there's only one line with a cursor
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/LazyVim/LazyVim/discussions/843#discussioncomment-11362693
>You are receiving this because you commented.
>
>Message ID: ***@***.***>
bogdan bozic

**Comment by gybo03** [comment]
It shoud look like you only type on one line but again when you make a change to that one line and press escape it will make the same text in all the other lines you selected during block-v mode

On November 25, 2024 8:27:55 AM GMT+01:00, bogdan ***@***.***> wrote:
>Yes it looks like it but if you make a change in that line and then press Esc to escape insert mode it will apply to all previous selected lines
>
>On November 24, 2024 12:57:14 PM GMT+01:00, ling ***@***.***> wrote:
>>> i think this is the core functionality of Vim so by transitivity it is in LazyVim also, you dont need to install anything
>>
>>Strangely, I can't do it. When I press I, there's only one line with a cursor
>>
>>-- 
>>Reply to this email directly or view it on GitHub:
>>https://github.com/LazyVim/LazyVim/discussions/843#discussioncomment-11362693
>>You are receiving this because you commented.
>>
>>Message ID: ***@***.***>
>bogdan bozic
bogdan bozic

**Reply by LingLambda** [reply]
> It shoud look like you only type on one line but again when you make a change to that one line and press escape it will make the same text in all the other lines you selected during block-v mode
> 看起来您只在一行上输入了文字，但当您对这一行进行修改并按下 escape 键时，您在块-V 模式下选择的所有其他行上的文字都会相同。
> On November 25, 2024 8:27:55 AM GMT+01:00, bogdan ***@***.***> wrote: Yes it looks like it but if you make a change in that line and then press Esc to escape insert mode it will apply to all previous selected lines On November 24, 2024 12:57:14 PM GMT+01:00, ling ***@***.***> wrote: >> i think this is the core functionality of Vim so by transitivity it is in LazyVim also, you dont need to install anything > >Strangely, I can't do it. When I press I, there's only one line with a cursor > >-- >Reply to this email directly or view it on GitHub: >[#843 (reply in thread)](https://github.com/LazyVim/LazyVim/discussions/843#discussioncomment-11362693) >You are receiving this because you commented. > >Message ID: ***@***.***> bogdan bozic
> 2024 年 11 月 25 日上午 8:27:55 GMT+01:00, bogdan ***@****.***> 写道：
> 是的，看起来是这样，但如果你对该行进行了修改，然后按 Esc 键退出插入模式，它就会应用到之前选定的所有行。
> 
> 2024 年 11 月 24 日下午 12:57:14 GMT+01:00, ling ***@****.***> 写道：
> >> 我认为这是 Vim 的核心功能，所以 LazyVim 中也有，你不需要安装任何东西。
> >
> 很奇怪，我做不到。当我按 I 键时，只有一行有光标
> >
> >--
> 请直接回复此邮件或在 GitHub 上查看：
> > [#843 (reply in thread)](https://github.com/LazyVim/LazyVim/discussions/843#discussioncomment-11362693) 
> >您收到此信息是因为您发表了评论。
> >
> >信息 ID：****@****.****>
> 博格丹-博齐奇
> bogdan bozic 博格丹-博齐奇

Thank you for your help, it works.

**Comment by gybo03** [comment]
Im glad to hear it

On November 29, 2024 8:58:11 AM GMT+01:00, ling ***@***.***> wrote:
>> It shoud look like you only type on one line but again when you make a change to that one line and press escape it will make the same text in all the other lines you selected during block-v mode
>> 看起来您只在一行上输入了文字，但当您对这一行进行修改并按下 escape 键时，您在块-V 模式下选择的所有其他行上的文字都会相同。
>> On November 25, 2024 8:27:55 AM GMT+01:00, bogdan ***@***.***> wrote: Yes it looks like it but if you make a change in that line and then press Esc to escape insert mode it will apply to all previous selected lines On November 24, 2024 12:57:14 PM GMT+01:00, ling ***@***.***> wrote: >> i think this is the core functionality of Vim so by transitivity it is in LazyVim also, you dont need to install anything > >Strangely, I can't do it. When I press I, there's only one line with a cursor > >-- >Reply to this email directly or view it on GitHub: >[#843 (reply in thread)](https://github.com/LazyVim/LazyVim/discussions/843#discussioncomment-11362693) >You are receiving this because you commented. > >Message ID: ***@***.***> bogdan bozic
>> 2024 年 11 月 25 日上午 8:27:55 GMT+01:00, bogdan ***@****.***> 写道：
>> 是的，看起来是这样，但如果你对该行进行了修改，然后按 Esc 键退出插入模式，它就会应用到之前选定的所有行。
>> 
>> 2024 年 11 月 24 日下午 12:57:14 GMT+01:00, ling ***@****.***> 写道：
>> >> 我认为这是 Vim 的核心功能，所以 LazyVim 中也有，你不需要安装任何东西。
>> >
>> 很奇怪，我做不到。当我按 I 键时，只有一行有光标
>> >
>> >--
>> 请直接回复此邮件或在 GitHub 上查看：
>> > [#843 (reply in thread)](https://github.com/LazyVim/LazyVim/discussions/843#discussioncomment-11362693) 
>> >您收到此信息是因为您发表了评论。
>> >
>> >信息 ID：****@****.****>
>> 博格丹-博齐奇
>> bogdan bozic 博格丹-博齐奇
>
>Thank you for your help, it works.
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/LazyVim/LazyVim/discussions/843#discussioncomment-11413277
>You are receiving this because you commented.
>
>Message ID: ***@***.***>
bogdan bozic

**Comment by michelaquino** [comment]
Just add this in some `.lua` file in the `plugins` directory:
```
return {
  "mg979/vim-visual-multi",
  branch = "master",
  init = function()
    vim.g.VM_maps = {
      ["Find Under"] = "<C-d>",
      ["Find Subword Under"] = "<C-d>",
      ["Select All"] = "<C-a>",
      ["Add Cursor Down"] = "<C-Down>",
      ["Add Cursor Up"] = "<C-Up>",
    }
    vim.g.VM_leader = "\\"
  end,
}

```
